File: clj/p3.clj
;;Q: What is the largest prime factor of the number 600851475143?
;;A: 6857

(defn is-prime [n]
  (let [n (int n)]
    (if (zero? (rem n (int 2)))
      false
      (not-any? zero? (map rem
			   (repeat n)
			   (take-nth 2 (range 3 (+ (. Math sqrt n) 1))))))))

(defn largest-factor []
  (loop [num 600851475143, max (+ (/ num 2) 1), i 3, current 1]
    (if (= current num)
      (- i 2)
      (if (and (is-prime i) (zero? (rem num i)))
	(recur num max (+ i 2) (* current i))
	(recur num max (+ i 2) current)))))

File: python/p3.py
#Q: What is the largest prime factor of the number 600851475143?
#A: 6857

def isprime(n):
    i=3
    if n%2 == 0:
        return False
    while i < (n**0.5)+1:
        if n%i == 0:
            return False
        i += 2
    return True

num = 600851475143
max = num//2+1
i = 3
current = 1
while i < max:
    if isprime(i) and num%i==0:
        current = current*i
        if current == num:
            print(str(i))
            break
    i += 2
